### PR TITLE
feat(crawl-hash-urls): Enable crawling for URLs having hash fragments in crawler

### PR DESCRIPTION
**Pull request by v-sharmachir**
#### Details

Added new optional parameter keepUrlFragment to enable crawling for URLs having hash fragments.
The default value of it is false so that existing scan work as is.

Added a test application with accessibility issues in the dev folder for end to end testing. Later on, same application will be used in release validation pipelines.

##### Motivation

[Feature 2114928](https://dev.azure.com/mseng/1ES/_workitems/edit/2114928): ADO Extension should be able to crawl hash-routing SPA apps

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Added relevant unit test for your changes. (`yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: dev/website-hashroute/src/Linked1/InnerPage.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import React from 'react';
+import { NavLink } from 'react-router-dom';
+
+const InnerPage = () => {
```
Comment (v-rakeshsh) [inline dev/website-hashroute/src/Linked1/InnerPage.js:7]: please add comments about InnerPage above const InnerPage, it should be self explanatory.
Same for all wherever a new component is added.

Comment (v-sharmachir) [reply dev/website-hashroute/src/Linked1/InnerPage.js:8]: added the description

File: docs/ado-extension-inputs.md [hunk 2]
```diff
@@ -28,6 +28,8 @@ or
 
 -   `staticSitePort` (int). If using `staticSiteDir`, you can set the local website TCP port for the extension to use when serving the static site. If unspecified, a port will be set automatically. If the pipeline serves additional processes, set the port to avoid conflicts. If using `baselineFile`, set the port to make sure future scans will match the baseline file.
 
+-   `keepUrlFragment` (boolean). To keep the hash fragment in the URLs. If set to false, it will remove the hash fragment from URL for example, http://www.example.com/#foo will be considered as http://www.example.com.
```
Comment (v-viyada) [inline docs/ado-extension-inputs.md:31]: mention the default value
`keepUrlFragment` (boolean) Default: `false`.

File: packages/ado-extension/e2e/ado-extension.test.ts [hunk 3]
```diff
@@ -203,6 +203,28 @@ describe('Sample task tests', () => {
...
+        };
+        const testSubject = runTestWithInputs(inputs);
+
+        expect(testSubject.stdOutContained('Accessibility scanning of URL http://localhost:3000/ completed')).toBeTruthy();
+        expect(testSubject.stdOutContainedRegex(new RegExp('Navigate page to URL.*{"url":"http://localhost:3000/#linked1"}'))).toBeTruthy();
+    });
+
+    it('should not crawl hash url with `keepUrlFragment` as false', () => {
+        inputs = {
+            url: 'http://127.0.0.1:5500/',
```
Comment (v-viyada) [inline packages/ado-extension/e2e/ado-extension.test.ts:219]: Url should be http://localhost:3000/

Comment (v-sharmachir) [reply packages/ado-extension/e2e/ado-extension.test.ts]: resolved

File: docs/ado-extension-inputs.md (hunk 2 above)
Comment (v-sharmachir) [reply docs/ado-extension-inputs.md]: resolved

Comment (v-sharmachir) [reply docs/ado-extension-inputs.md]: resolved

File: packages/ado-extension/package.json [hunk 4]
```diff
@@ -14,7 +14,7 @@
         "start": "node scripts/run-locally.js",
         "test": "jest --rootDir src",
         "watch:test": "jest --watch",
-        "test:e2e": "jest --rootDir e2e --coverage false",
+        "test:e2e": "concurrently \"npm --prefix ../test-website-hashroute start\" \"jest --rootDir e2e --coverage false\" --kill-others --success first",
```
Comment (madalynrose) [inline packages/ado-extension/package.json:17]: We can avoid pulling in an additional library by using `npm-run-all` and splitting this into two commands:
```suggestion
        "test:e2e": "npm-run-all --parallel test:e2e:serve-hashroute test:e2e:run-tests",
        "test:e2e:run-tests": "jest --rootDir e2e --coverage false",
        "test:e2e:serve-hashroute": "yarn workspace test-website-hashroute start",
```

File: packages/ado-extension/package.json [hunk 5]
```diff
@@ -37,6 +37,7 @@
         "@typescript-eslint/eslint-plugin": "^6.18.1",
         "@typescript-eslint/parser": "^6.18.1",
         "case-sensitive-paths-webpack-plugin": "^2.4.0",
+        "concurrently": "^8.2.2",
```
Comment (madalynrose) [inline packages/ado-extension/package.json:40]: We can remove this after updating the `test:e2e` script
```suggestion
```

**Review by madalynrose — CHANGES_REQUESTED** [review]
I've got one adjustment to the script for e2e tests but overall this looks great!

File: packages/ado-extension/package.json (hunk 4 above)
Comment (v-viyada) [reply packages/ado-extension/package.json]: Removed concurrently and used npm-run-all

File: packages/ado-extension/package.json (hunk 5 above)
Comment (v-viyada) [reply packages/ado-extension/package.json]: Removed

**Review by madalynrose — APPROVED** [review]
This looks awesome!!

**Pull request merged**